File: 2025/01/main.rb
pos = 50
number_of_zeroes = 0

File.read('input.txt').split("\n").each do |cmd|
  direction = cmd.slice 0, 1
  distance = cmd[1..-1].to_i
  case direction
  when 'L'
    pos -= distance
  when 'R'
    pos += distance
  else
    raise "Unreachable: #{direction}"
  end
  pos = pos % 100
  puts "After #{cmd} position is now #{pos}"
  if pos == 0
    number_of_zeroes += 1
  end
end

puts "Number of zeroes: #{number_of_zeroes}"

File: 2025/01/main.ml
#!/usr/bin/env ocaml

let () =
  let lines = In_channel.open_text "input.txt" |> In_channel.input_lines in
  let (_, zeroes) =
    List.fold_left
      (fun (pos, zeroes) line ->
        let direction = String.sub line 0 1 in
        let distance =
          String.sub line 1 (String.length line - 1) |> int_of_string
        in
        let pos =
          match direction with
          | "L" -> pos - distance
          | "R" -> pos + distance
          | _ -> failwith "Unreachable"
        in
        let pos = pos mod 100 in
        let zeroes = if pos = 0 then zeroes + 1 else zeroes in
        (pos, zeroes))
      (50, 0) lines
  in
  Printf.printf "Number of zeroes: %d\n" zeroes
